appengine-jetty-managed-runtime/src/main/java/com/google/apphosting/vmruntime/jetty9/VmRuntimeWebAppContext.java

```diff
@@ -262,129 +260,97 @@ public boolean isTrustedRemoteAddr(String remoteAddr) {
+      VmApiProxyEnvironment.createFromHeaders(
+          System.getenv(), metadataCache, this, VmRuntimeUtils.getApiServerAddress(),
+          wallclockTimer, VmRuntimeUtils.ONE_DAY_IN_MILLIS, defaultEnvironment);
     }
 
+    VmApiProxyEnvironment getRequestSpecificEnvironment() {
+      return requestSpecificEnvironment;
+    }
+  }
+  
+  private class ContextListener implements ContextHandler.ContextScopeListener, ServletRequestListener {
     @Override
-    public void log(String message) {
-      log(message, null);
+    public void enterScope(org.eclipse.jetty.server.handler.ContextHandler.Context context, Request baseRequest, Object reason) {
+      RequestContext requestContext = getRequestContext(baseRequest);
+      if (LOG.isDebugEnabled())
+        LOG.debug("Enter {} -> {}",ApiProxy.getCurrentEnvironment(),requestContext.getRequestSpecificEnvironment());
+      ApiProxy.setEnvironmentForCurrentThread(requestContext.getRequestSpecificEnvironment());      
     }
 
-    /**
-     * {@inheritDoc}
-     *
-     * @param throwable an exception associated with this log message,
-     * or {@code null}.
-     */
     @Override
-    public void log(String message, Throwable throwable) {
-      StringWriter writer = new StringWriter();
-      writer.append("javax.servlet.ServletContext log: ");
-      writer.append(message);
+    public void requestInitialized(ServletRequestEvent sre) {
+      ServletRequest request = sre.getServletRequest();
+      Request baseRequest = Request.getBaseRequest(request);
+      RequestContext requestContext = getRequestContext(baseRequest);
+      
+      // Check for SkipAdminCheck and set attributes accordingly.
+      VmRuntimeUtils.handleSkipAdminCheck(requestContext);
 
-      if (throwable != null) {
-        writer.append("\n");
-        throwable.printStackTrace(new PrintWriter(writer));
+      // Change scheme to HTTPS based on headers set by the appserver.
+      setSchemeAndPort(baseRequest);
+    }
+    
+    @Override
+    public void requestDestroyed(ServletRequestEvent sre) {      
+      ServletRequest request = sre.getServletRequest();
+      Request baseRequest = Request.getBaseRequest(request);
+      RequestContext requestContext = getRequestContext(baseRequest);
+      VmApiProxyEnvironment env = requestContext.getRequestSpecificEnvironment();
+      
+      // TODO is this interrupting and waiting still needed?
+      if (request.isAsyncStarted()) {
+        request.getAsyncContext().addListener(new AsyncListener() {
+          @Override public void onTimeout(AsyncEvent event) throws IOException {}
+          @Override public void onStartAsync(AsyncEvent event) throws IOException {}
+          @Override public void onError(AsyncEvent event) throws IOException {}
+          
+          @Override
+          public void onComplete(AsyncEvent event) throws IOException {
+            VmRuntimeUtils.interruptRequestThreads(env, VmRuntimeUtils.MAX_REQUEST_THREAD_INTERRUPT_WAIT_TIME_MS);
```

Comment:
@ludoch is it still required to interrupt and wait for all API calls at the completion of each request?